grr/server/grr_response_server/gui/api_plugins/metadata.py

```diff
@@ -51,14 +60,546 @@ class ApiGetOpenApiDescriptionHandler(api_call_handler_base.ApiCallHandler):
+      protobuf2.TYPE_FLOAT: "protobuf2.TYPE_FLOAT",
+      protobuf2.TYPE_INT64: "protobuf2.TYPE_INT64",
+      protobuf2.TYPE_UINT64: "protobuf2.TYPE_UINT64",
+      protobuf2.TYPE_INT32: "protobuf2.TYPE_INT32",
+      protobuf2.TYPE_FIXED64: "protobuf2.TYPE_FIXED64",
+      protobuf2.TYPE_FIXED32: "protobuf2.TYPE_FIXED32",
+      protobuf2.TYPE_BOOL: "protobuf2.TYPE_BOOL",
+      protobuf2.TYPE_STRING: "protobuf2.TYPE_STRING",
+      protobuf2.TYPE_BYTES: "protobuf2.TYPE_BYTES",
+      protobuf2.TYPE_UINT32: "protobuf2.TYPE_UINT32",
+      protobuf2.TYPE_SFIXED32: "protobuf2.TYPE_SFIXED32",
+      protobuf2.TYPE_SFIXED64: "protobuf2.TYPE_SFIXED64",
+      protobuf2.TYPE_SINT32: "protobuf2.TYPE_SINT32",
+      protobuf2.TYPE_SINT64: "protobuf2.TYPE_SINT64",
+    }
+    self.primitive_types_names = \
+      list(self.proto_primitive_types_names.values()) \
+      + ["BinaryStream",]
+
+  def _SimplifyPathNode(self, node: str) -> str:
+    if len(node) > 0 and node[0] == '<' and node[-1] == '>':
+      node = node[1:-1]
+      node = node.split(":")[-1]
+      node = f"{{{node}}}"
+
+    return node
+
+  def _SimplifyPath(self, path: str) -> str:
+    """Keep only fixed parts and parameter names from Werkzeug URL patterns.
+
+    The OpenAPI specification requires that parameters are surrounded by { }
+    which are added in _SimplifyPathNode.
+    """
+
+    nodes = path.split("/")
+    simple_nodes = [self._SimplifyPathNode(node) for node in nodes]
+
+    simple_path = '/'.join(simple_nodes)
+
+    return simple_path
+
+  def _GetPathArgsFromPath(self, path: str) -> [str]:
+    """Extract path parameters from a Werkzeug Rule URL."""
+    path_args = []
+
+    nodes = path.split("/")
+    for node in nodes:
+      if len(node) > 0 and node[0] == '<' and node[-1] == '>':
+        simple_node = self._SimplifyPathNode(node)
+        simple_node = simple_node[1:-1]
+        path_args.append(simple_node)
+
+    return path_args
+
+  def _GetTypeName(self, t):
+    if isinstance(t, FieldDescriptor):
+      if t.message_type:
+        return self._GetTypeName(t.message_type)
+      if t.enum_type:
+        return self._GetTypeName(t.enum_type)
+
+      return self._GetTypeName(t.type)
+
+    if isinstance(t, Descriptor):
+      return t.full_name
+
+    if isinstance(t, EnumDescriptor):
+      return t.full_name
+
+    if inspect.isclass(t):
+      return t.__name__
+
+    if isinstance(t, int): # It's a protobuf.Descriptor.type value.
+      return self.proto_primitive_types_names[t]
+
+    return str(t) # Cover "BinaryStream" and None.
+
+  def _SetMetadata(self):
+    oas_version = "3.0.3"
+    self.open_api_obj["openapi"] = oas_version
+
+    # The Info Object "info" field.
+    info_obj = dict()
+    info_obj["title"] = "GRR Rapid Response API"
+    info_obj["description"] = "GRR Rapid Response is an incident response " \
+                              "framework focused on remote live forensics."
```

Comment:
Why not use dictionary literals (here and below)? Also, since this is mostly static representation, maybe using a global constant for this is a better idea (something to discuss).

Reply:
Thank you for pointing this out! I made the move to dictionary literals where I thought it was possible. Please let me know if you find other instances that make sense to be converted to dictionary literals.

Regarding the `info_obj`, I think it might be useful to keep it not set in stone in order to allow customization for special cases, if such special cases can exist (?). For example, would it be possible to want different info objects for different `routers` ? In this case we could override the associated method in subclasses that get instantiated by different routers' methods. Or we could use a static variable per subclass (this might require metaclasses) to replace the overriding ooor, we could pass the `info_obj` to the constructor of the `Handler` from the router's method. What do you think?